### PR TITLE
fix(gateway): include CORS on subdomain redirects

### PR DESCRIPTION
**Pull request by lidel**
This PR closes regression introduced during one of the refactors since Kubo 0.20 and adds dedicated tests for preflight requests on both path and subdomain gateway.

Context: https://github.com/ipfs/kubo/issues/9983#issuecomment-1599673976

### CODE REVIEW

File: gateway/gateway.go
Comment (lidel) [inline gateway/gateway.go:356]: :information_source:  we were missing HEAD and OPTIONS, which may not be used a lot today, but will be once HTTP transport becomes more and more popular. Included tests for this too.

File: gateway/gateway_test.go
Comment (lidel) [inline gateway/gateway_test.go]: :information_source: basic regression test that guards CORS-related headers on path and subdomain gateways

File: gateway/gateway.go
Comment (lidel) [inline gateway/gateway.go:339]: @laurentsenta  this is just FYI if you ever need to find where `boxo/gateway` sets default CORS headers :)

Once this ships in Kubo, we can update conformance [TestCors](https://github.com/ipfs/gateway-conformance/blob/43c0842a1041a2748a74a196e370b84ef02b12a5/tests/path_gateway_unixfs_test.go#L455) to include all headers set by `AddAccessControlHeaders` here.

I've also filled https://github.com/ipfs/specs/issues/423 to write this down and add it to specs (currently CORS on gateways is grossly undocumented), but fine to update conformance before specs land.

**Review by hacdias — APPROVED** [review]
LGTM, I rebased it. Will merge once CI is green.

**Pull request merged**
